File: Leet 25/solution.py
class ListNode:
    def __init__(self, val: int=0):
        self.val: int = val
        self.next: ListNode|None = None


class Solution:
    def list_to_listnode(self, values: list[int]) -> ListNode|None:
        head: ListNode|None = None
        tail: ListNode|None = None

        for val in values:
            if head == None:
                head = ListNode(val)
                tail = head
            elif tail:
                tail.next = ListNode(val)
                tail = tail.next

        return head

    def reverseKGroup(self, head: ListNode|None, k: int) -> ListNode|None:
        if not head:
            return None
        
        values: list[int] = []
        curr: ListNode|None = head
        while curr:
            values.append(curr.val)
            curr = curr.next

        n: int = len(values)
        for i in range(0, n, k):
            if i + k <= n:
                for j, v in enumerate(reversed(values[i:i+k])):
                    values[i + j] = v

        return self.list_to_listnode(values)


def print_listnode(head: ListNode|None) -> None:
    if not head:
        return
    
    current = head
    print('[', end='')
    while current:
        print(f' {current.val} ', end='')
        current = current.next
    print(']')


def test_case(solution: Solution, nums: list[int], k: int) -> None:
    head = solution.list_to_listnode(nums)
    head = solution.reverseKGroup(head, k)
    print_listnode(head)


def main() -> None:
    print('25. Reverse Nodes in k-Group')

    solution = Solution()

    test_case(solution, [1,2,3,4,5], 2)
    test_case(solution, [1,2,3,4,5], 3)
    test_case(solution, [1,2], 2)


if __name__ == '__main__':
    main()

File: Leet 25/solution.java
public class ListNode 
{
    int val;
    ListNode next;
    
    ListNode(int val) 
    { 
        this.val = val; 
    }
}

public class Solution 
{
    private ListNode fromArrayList(ArrayList<Integer> values) 
    {
        ListNode head = null;
        ListNode tail = null;

        for (var value : values) 
        {
            if (head == null) 
            {
                head = new ListNode(value);
                tail = head;
            } 
            else 
            {
                tail.next = new ListNode(value);
                tail = tail.next;
            }
        }
        return head;
    }

    public ListNode reverseKGroup(ListNode head, int k) 
    {
        if (head == null) 
        {
            return null;
        }

        var values = new ArrayList<Integer>();
        var current = head;
        while (current != null) 
        {
            values.add(current.val);
            current = current.next;
        }

        for(int i = 0; i < values.size(); i += k)
        {
            if (i + k <= values.size())
            {
                var sub = values.subList(i, i + k);
                Collections.reverse(sub);

                for(int j = 0; j < sub.size(); j++)
                {
                    values.set(i + j, sub.get(j));
                }
            }
        }

        return fromArrayList(values);
    }
}